adapters/coap-vertx-quarkus/src/main/java/org/eclipse/hono/adapter/coap/InternalErrorTracer.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.adapter.coap;
+
+import java.util.Optional;
+
+import org.eclipse.californium.core.coap.Response;
+import org.eclipse.californium.core.network.interceptors.MessageInterceptorAdapter;
+import org.eclipse.californium.core.network.stack.BlockwiseLayer;
+import org.eclipse.californium.elements.EndpointContext;
+import org.eclipse.californium.elements.auth.AbstractExtensiblePrincipal;
+import org.eclipse.californium.elements.auth.AdditionalInfo;
+import org.eclipse.hono.auth.Device;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+
+/**
+ * There are some errors in the CoAP protocol that occur on a lower level, for example {@link BlockwiseLayer}.
+ * Californium returns an error response to the client but this is not propagated to the Hono CoAP adapter. This is
+ * the purpose of this interceptor - to track if such errors occur and report a trace for them in Jaeger.
+ */
+public class InternalErrorTracer extends MessageInterceptorAdapter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(InternalErrorTracer.class);
+
+    private final String coapAdapterName;
+    private final Tracer tracer;
+
+    /**
+     * @param coapAdapterName Hono CoAP adapter component name.
+     * @param tracer Open Tracing tracer to use.
+     */
+    public InternalErrorTracer(final String coapAdapterName, final Tracer tracer) {
+        this.coapAdapterName = coapAdapterName;
+        this.tracer = tracer;
+    }
+
+    @Override
+    public void sendResponse(final Response response) {
+        if (response.isInternal() && response.isError()) {
+            final EndpointContext context = response.getDestinationContext();
+            final AdditionalInfo honoInfo = Optional.ofNullable(context.getPeerIdentity())
+                    .filter(AbstractExtensiblePrincipal.class::isInstance)
+                    .map(AbstractExtensiblePrincipal.class::cast)
+                    .map(AbstractExtensiblePrincipal::getExtendedInfo)
+                    .orElse(null);
+            if (honoInfo == null) {
+                LOG.warn("CoAP internal error has occured, but Hono context info is not set in the response: "
```

Comment:
IMHO it should be ok to simply log this at DEBUG level as this does not pose any problem, does it?
However, please use the string formatting mechanism of the the Logger instead of String concatenation.

Reply:
Well, it would be strange if Hono context is not set.. First because it must be (as far as I understood) and second we cannot report a trace in that case, even though there's a CoAP internal error. If there's no trace, would the debug log catch our attention enough?

Reply:
I see your point. But I guess we could still create a reasonable span based on the information from the request that this is the response to. The addition info being `null` means that the device has not been authenticated. However, we can still create a span that for example just contains the request URI and the response code.

Reply:
Ok, so if the additional info being `null` is a valid case, all we can log from the response is device IP/port, error code, some security info and the payload. We don't have any info about the request as far as I see. Will it bring value to create a span with IP/port when they can change at any time?

If not, we should think how to handle the case with unauthenticated devices... Maybe skip them at all?

Reply:
> We don't have any info about the request as far as I see. 

I guess you are right. I had originally thought that we would also have a reference to the corresponding Request, but that is only available in the CoapExchange.

> If not, we should think how to handle the case with unauthenticated devices... Maybe skip them at all?

There is nothing in particular that we can do about it. Logging this at WARN level does not really help anybody remedying the situation as this is clearly a client/device error, isn't it? So, FMPOV, we should simply log it at DEBUG level and that's it ...